File: src/cocktailDB.R
#!/usr/bin/env Rscript
# This script was created for a technical exercise.
# The data used is from TheCocktailDB API. This is a free to use dataset
# This script will process the json data from the api normalise and standardise the output
# and outputs both raw and transformed data to files.
# These files can be used for further analysis and reporting using the transformed data 
# or verification using the original raw data.

# Define a vector of required packages
required_packages <- c("httr", "jsonlite", "stringr")

# Function to check and install missing packages
install_if_missing <- function(packages) {
  for (pkg in packages) {
    if (!require(pkg, character.only = TRUE)) {
      install.packages(pkg, dependencies = TRUE)
    }
  }
}

# Install missing packages
install_if_missing(required_packages)

library(httr)
library(jsonlite)
library(stringr)

# 1 Data Ingestion
fetch_cocktail_data <- function() {
  url <- "https://www.thecocktaildb.com/api/json/v1/1/random.php"
  response <- GET(url)

  if (status_code(response) == 200) {
    data <- content(response, as = "parsed")
    return(data$drinks[[1]])
  } else {
    stop(sprintf("Failed to fetch data. Status code: %d", status_code(response)))
  }
}

standardise_units <- function(measures) {
  unit_conversion <- list(
    oz = 29.57,
    ml = 1.0,
    cl = 10.0,
    tsp = 4.93,
    tbsp = 14.79
  )

  standardised_measures <- sapply(measures, function(measure) {
    if (!is.null(measure)) {
      match <- str_match(measure, "(\\d+(\\.\\d+)?)\\s*(\\w+)")
      if (!is.na(match[1])) {
        quantity <- as.numeric(match[2])
        unit <- tolower(match[4])
        if (unit %in% names(unit_conversion)) {
          standardised_quantity <- quantity * unit_conversion[[unit]]
          return(sprintf("%.2f ml", standardised_quantity))
        } else {
          return(measure)  # No change if unknown unit
        }
      } else {
        return(measure)  # No change if no match
      }
    } else {
      return(NA)
    }
  })

  return(standardised_measures)
}

normalise_string <- function(value) {
  if (!is.null(value)) {
    return(trimws(gsub("[^\\w\\s]", "", value)))
  }
  return(value)
}

# 2. Data transformation inc. standardisation and normalisation
transform_cocktail_data <- function(raw_data) {
  ingredients <- unlist(lapply(1:15, function(i) raw_data[[paste0("strIngredient", i)]]))
  ingredients <- ingredients[!is.null(ingredients)]

  measures <- unlist(lapply(1:15, function(i) raw_data[[paste0("strMeasure", i)]]))
  measures <- measures[!is.null(measures)]

  return(list(
    CocktailID = raw_data$idDrink,
    Name = normalise_string(raw_data$strDrink),
    Category = raw_data$strCategory,
    Alcoholic = tolower(raw_data$strAlcoholic) == "alcoholic",
    GlassType = raw_data$strGlass,
    Ingredients = ingredients,
    Measures = standardise_units(measures),
    Instructions = normalise_string(raw_data$strInstructions)
  ))
}

# 3. Edge cases
handle_edge_cases <- function(transformed_data) {
  ingredients <- transformed_data$Ingredients
  measures <- transformed_data$Measures
  
  # Ensure both vectors are of equal length
  paired_data <- Filter(
    function(x) !is.null(x[[1]]) && !is.null(x[[2]]),
    mapply(
      list,
      head(ingredients, min(length(ingredients), length(measures))),
      head(measures, min(length(ingredients), length(measures))),
      SIMPLIFY = FALSE
    )
  )
  
  transformed_data$Ingredients <- sapply(paired_data, `[[`, 1)
  transformed_data$Measures <- sapply(paired_data, `[[`, 2)
  
  return(transformed_data)
}

main <- function(batch_size = 10) {
  raw_data_batch <- list()
  transformed_data_batch <- list()

  for (i in 1:batch_size) {
    tryCatch({
      raw_data <- fetch_cocktail_data()
      raw_data_batch <- append(raw_data_batch, list(raw_data))

      transformed_data <- transform_cocktail_data(raw_data)
      transformed_data <- handle_edge_cases(transformed_data)
      transformed_data_batch <- append(transformed_data_batch, list(transformed_data))

    }, error = function(e) {
      message(sprintf("Error processing cocktail: %s", e$message))
    })
  }

  write(toJSON(raw_data_batch, pretty = TRUE, auto_unbox = TRUE), "raw_cocktail_data.json")
  write(toJSON(transformed_data_batch, pretty = TRUE, auto_unbox = TRUE), "transformed_cocktail_data.json")

  print("Batch processing complete. Data saved to files.")
}

if (interactive()) {
  main()
}

File: src/cocktailDB.py
#!/usr/bin/env python3
# This script was created for a technical exercise.
# The data used is from TheCocktailDB API. This is a free to use dataset
# This script will process the json data from the api normalise and standardise the output
# and outputs both raw and transformed data to files.
# These files can be used for further analysis and reporting using the transformed data 
# or verification using the original raw data.

import requests
import json
import re

# 1 Data Ingestion
def fetch_cocktail_data(): # Fetch data, exception status_code =! 200
    url = "https://www.thecocktaildb.com/api/json/v1/1/random.php"
    response = requests.get(url)
    if response.status_code == 200:
        return response.json()["drinks"][0]  # Extract the cocktail data
    else:
        raise Exception(f"Failed to fetch data. Status code: {response.status_code}")
    
def standardise_units(measures): # Standardise units into SI (ml)
    unit_conversion = {
        "oz": 29.57,
        "ml": 1.0,
        "cl": 10.0,
        "tsp": 4.93,
        "tbsp": 14.79,
    }
    standardised_measures = []

    for measure in measures:
        if measure:
            match = re.match(r"(\d+(\.\d+)?)\s*(\w+)", measure) # Match: one or more digit followed by optional decimal and one or more digit
            if match:
                quantity, _, unit = match.groups()
                quantity = float(quantity)
                unit = unit.lower()
                if unit in unit_conversion:
                    standardised_quantity = quantity * unit_conversion[unit]
                    standardised_measures.append(f"{standardised_quantity:.2f} ml")
                else:
                    standardised_measures.append(measure)  # No change if unknown unit
            else:
                standardised_measures.append(measure)  # No change if no match
        else:
            standardised_measures.append(None)

    return standardised_measures

def normalise_string(value): # Removes special char and white space char
    if value:
        return re.sub(r"[^\w\s]", "", value).strip() # Match: Not word char or white space char
    return value

# 2. Data transformation inc. standardisation and normalisation
def transform_cocktail_data(raw_data): # Custom mapping
    ingredients = [
        raw_data[f"strIngredient{i}"] for i in range(1, 16) if raw_data[f"strIngredient{i}"]
    ]
    measures = [
        raw_data[f"strMeasure{i}"] for i in range(1, 16) if raw_data[f"strMeasure{i}"]
    ]

    return {
        "CocktailID": raw_data["idDrink"],
        "Name": normalise_string(raw_data["strDrink"]),
        "Category": raw_data["strCategory"],
        "Alcoholic": raw_data["strAlcoholic"].lower() == "alcoholic",
        "GlassType": raw_data["strGlass"],
        "Ingredients": ingredients,
        "Measures": standardise_units(measures),
        "Instructions": normalise_string(raw_data["strInstructions"]),
    }

# 3. Edge cases
def handle_edge_cases(transformed_data): # Handle edge cases in the transformed data
    # Remove ingredients with no matching measurements and vice versa
    ingredients = transformed_data["Ingredients"]
    measures = transformed_data["Measures"]

    paired_data = [
        (ingredient, measure)
        for ingredient, measure in zip(ingredients, measures)
        if ingredient and measure
    ]

    transformed_data["Ingredients"] = [pair[0] for pair in paired_data]
    transformed_data["Measures"] = [pair[1] for pair in paired_data]

    return transformed_data

def main(batch_size=10):
    raw_data_batch = []
    transformed_data_batch = []

    for _ in range(batch_size):
        try:
            raw_data = fetch_cocktail_data()
            raw_data_batch.append(raw_data)

            transformed_data = transform_cocktail_data(raw_data)
            transformed_data = handle_edge_cases(transformed_data)
            transformed_data_batch.append(transformed_data)

        except Exception as e:
            print(f"Error processing cocktail: {e}")

    # Save raw and transformed if needed from batch processing.
    with open("raw_cocktail_data.json", "w") as raw_file:
        json.dump(raw_data_batch, raw_file, indent=4)

    with open("transformed_cocktail_data.json", "w") as transformed_file:
        json.dump(transformed_data_batch, transformed_file, indent=4)

    print("Batch processing complete. Data saved to files.")

if __name__ == "__main__":
    main()
